crates/ruff_python_formatter/src/verbatim.rs

```diff
@@ -0,0 +1,505 @@
+                        ]
+                    )?;
+
+                    format_on_comment.mark_formatted();
+
+                    if let Some(new_format_off_comment) = new_format_off_comment {
+                        format_off_comment = new_format_off_comment;
+                        format_comment(format_off_comment).fmt(f)?;
+                        format_off_comment.mark_formatted();
+                    } else {
+                        return Ok(statement.statement());
+                    }
+                }
+
+                // Unreachable because the function exits as soon as it reaches the end of the suppression
+                // and it already starts in a suppressed range.
+                SuppressionComments::SuppressionStarts { .. } => unreachable!(),
+                SuppressionComments::Formatted { .. } => unreachable!(),
+            }
+        }
+
+        if let Some(next_statement) = statements.next() {
+            statement = SuiteChildStatement::Other(next_statement);
+            leading_node_comments = comments.leading_comments(next_statement);
+        } else {
+            let end = comments
+                .trailing_comments(statement)
+                .last()
+                .map_or(statement.end(), Ranged::end);
+
+            verbatim_text(TextRange::new(format_off_comment.end(), end)).fmt(f)?;
+
+            return Ok(statement.statement());
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+enum InSuppression {
+    No,
+    Yes,
+}
+
+#[derive(Debug)]
+enum SuppressionComments<'a> {
+    /// The first `fmt: off` comment.
+    SuppressionStarts {
+        /// The comments appearing before the `fmt: off` comment
+        formatted_comments: &'a [SourceComment],
+        format_off_comment: &'a SourceComment,
+    },
+
+    /// A `fmt: on` comment inside a suppressed range.
+    SuppressionEnds {
+        /// The comments before the `fmt: on` comment that should *not* be formatted.
+        suppressed_comments: &'a [SourceComment],
+        format_on_comment: &'a SourceComment,
+
+        /// The comments after the `fmt: on` comment (if any), that should be formatted.
+        formatted_comments: &'a [SourceComment],
+
+        /// Any following `fmt: off` comment if any.
+        /// * `None`: The suppression ends here (for good)
+        /// * `Some`: A `fmt: off`..`fmt: on` .. `fmt: off` sequence. The suppression continues after
+        ///     the `fmt: off` comment.
+        format_off_comment: Option<&'a SourceComment>,
+    },
+
+    /// Comments that all fall into the suppressed range.
+    Suppressed { comments: &'a [SourceComment] },
+
+    /// Comments that all fall into the formatted range.
+    Formatted {
+        #[allow(unused)]
+        comments: &'a [SourceComment],
+    },
+}
+
+impl<'a> SuppressionComments<'a> {
+    fn unwrap_suppression_starts(&self) -> (&'a [SourceComment], &'a SourceComment) {
+        match self {
```

Comment:
Nit: `if-let` for single-case match?

Reply:
I'll have a look. I love match because  I find it easier to reason about than `let...else` or `if let... else` and our pedantic rule annoys me a lot that forces me to rewrite the if else to a match